src/compiler/sys.ts

```diff
@@ -936,12 +980,67 @@ export function createSystemWatchFunctions({
     let nonPollingWatchFile: HostWatchFile | undefined;
     let hostRecursiveDirectoryWatcher: HostWatchDirectory | undefined;
     let hitSystemWatcherLimit = false;
+    let reportErrorOnMissingRequire = true;
+
     return {
         watchFile,
-        watchDirectory
+        watchDirectory,
     };
 
+    function getUserWatchFactory(options: WatchOptions | undefined): UserWatchFactory | undefined {
+        if (!options?.watchFactory) return undefined;
+        // Look if we alaready have this factory resolved
+        if (options.getResolvedWatchFactory) return options.getResolvedWatchFactory();
+        const system = getSystem();
+        if (!system.require) {
+            if (reportErrorOnMissingRequire) sysLog(`Custom watchFactory is ignored because of not running in environment that supports 'require'. Watches will defualt to builtin.`);
+            reportErrorOnMissingRequire = false;
+            return setUserWatchFactory(options, /*userWatchFactory*/ undefined);
+        }
+        const factoryName = isString(options.watchFactory) ? options.watchFactory : options.watchFactory.name;
+        if (!factoryName || parsePackageName(factoryName).rest) {
+            sysLog(`Skipped loading watchFactory ${isString(options.watchFactory) ? options.watchFactory : JSON.stringify(options.watchFactory)} because it can be named with only package name`);
+            return setUserWatchFactory(options, /*userWatchFactory*/ undefined);
+        }
+        const host = options.getHost?.();
+        const searchPaths = host ?
+            host.searchPaths :
+            [
+                combinePaths(system.getExecutingFilePath(), "../../..")
+            ];
+        sysLog(`Enabling watchFactory ${isString(options.watchFactory) ? options.watchFactory : JSON.stringify(options.watchFactory)} from candidate paths: ${searchPaths.join(",")}`);
+        const { resolvedModule, errorLogs, pluginConfigEntry } = resolveModule<UserWatchFactoryModule>(
+            getWatchFactoryPlugin(options),
+            searchPaths,
+            getSystem(),
+            sysLog
+        );
+        if (typeof resolvedModule === "function") {
+            return setUserWatchFactory(options, resolvedModule({ typescript: ts, options, config: pluginConfigEntry }));

```

Comment:
One complication here I just realized is that now that we are modules, this "ts namespace" object is only the view of the "namespace" from within the `compiler` project, meaning it won't include anything from `services`, `server`, `deprecatedCompat`, and so on. The latter one is probably the worst one because for perf reasons, we explicitly avoid including that in tsc.

We'll likely have to come up with some way to ensure the expected object makes it into the plugin, or restrict the API down to a limited subset.